File: frontend/src/types/agent.ts
export interface SubjectNode {
  name: string;
  source: 'user' | 'derived' | 'optional';
  children: string[];
  selected: boolean;
}

export interface CourseInfo {
  course_name: string;
  subject: string;
  department: string;
  grade_level: string;
}

export interface AgentConfig {
  description?: string;
  publishScope?: 'students' | 'teachers';
  major?: string;
  target_job?: string;
  core_skills?: string;
  certifications?: string;
  training_scenarios?: string;
  chunkSize?: number;
  chunkOverlap?: number;
  systemPrompt?: string;
  llmModel?: string;
  topK?: number;
  similarityThreshold?: number;
  role?: string;
  subject?: string;
  audience?: string;
  core_need?: string;
  style?: string;
  coreSubjects?: string[];
  knowledgeFileIds?: number[];
  teaching_style?: string;
  core_chapters?: string[];
  student_difficulties?: string;
  classroom_flow?: string;
  teaching_methods?: string[];
  interactive_methods?: string[];
  common_questions?: string[];
  assessment_method?: string;
  usual_ratio?: number;
  homework_types?: string[];
  textbook?: string;
  reference_books?: string[];
  teaching_tools?: string[];
  course_info?: CourseInfo;
  downloaded_from?: {
    agent_id: number;
    agent_name: string;
    author_name?: string;
    downloaded_at: string;
  };
}

export interface Agent {
  id: number;
  user_id: number;
  name: string;
  course_name: string;
  template: string;
  status: 'draft' | 'testing' | 'published';
  config: AgentConfig;
  version: number;
  created_at: string;
  updated_at: string | null;
}

export interface AgentCreateData {
  name: string;
  course_name: string;
  template?: string;
  config?: AgentConfig;
}

export interface AgentUpdateData {
  name?: string;
  course_name?: string;
  template?: string;
  config?: AgentConfig;
  status?: 'draft' | 'testing' | 'published';
}

export interface AgentMarketplaceItem {
  id: number;
  name: string;
  course_name: string;
  template: string;
  description?: string | null;
  subject?: string | null;
  department?: string | null;
  grade_level?: string | null;
  core_chapters: string[];
  teaching_tools: string[];
  llm_model?: string | null;
  version: number;
  created_at: string;
  updated_at: string | null;
  author_id: number;
  author_name?: string | null;
  author_department?: string | null;
  author_avatar?: string | null;
  usage_count: number;
  rating?: number | null;
  rating_count: number;
  config?: Record<string, unknown>;
}

export interface AgentMarketplacePage {
  items: AgentMarketplaceItem[];
  total: number;
  page: number;
  page_size: number;
}

File: backend/app/schemas/agent.py
import json
from pydantic import BaseModel, field_validator
from datetime import datetime


class AgentBase(BaseModel):
    name: str
    course_name: str


class AgentCreate(AgentBase):
    template: str = "higher_edu"
    config: dict = {}


class AgentUpdate(BaseModel):
    name: str | None = None
    course_name: str | None = None
    template: str | None = None
    status: str | None = None
    config: dict | None = None


class AgentResponse(AgentBase):
    id: int
    user_id: int
    template: str
    status: str
    config: dict
    version: int
    created_at: datetime
    updated_at: datetime | None

    class Config:
        from_attributes = True

    @field_validator("config", mode="before")
    @classmethod
    def parse_config(cls, v):
        if isinstance(v, str):
            return json.loads(v) if v else {}
        return v if isinstance(v, dict) else {}


class AgentMarketplaceItem(BaseModel):
    """市场列表中的单个 Agent 卡片数据。"""
    id: int
    name: str
    course_name: str
    template: str
    description: str | None = None
    subject: str | None = None
    department: str | None = None
    grade_level: str | None = None
    core_chapters: list[str] = []
    teaching_tools: list[str] = []
    llm_model: str | None = None
    version: int
    created_at: datetime
    updated_at: datetime | None
    # 作者信息
    author_id: int
    author_name: str | None = None
    author_department: str | None = None
    author_avatar: str | None = None
    # 统计（占位，后续可接 conversations/feedback 表聚合）
    usage_count: int = 0
    rating: float | None = None
    rating_count: int = 0
    config: dict = {}
    # 平台标记
    is_builtin: bool = False
    builtin_category: str | None = None  # 如："学科助教" / "教师备课" / "职业技能"

    @field_validator("description", mode="before")
    @classmethod
    def coerce_description(cls, v):
        """兼容数据库中 list 类型存储的 description。"""
        if isinstance(v, list):
            return " ".join(str(x) for x in v if x is not None)
        return v


class AgentMarketplacePage(BaseModel):
    """市场分页响应。"""
    items: list[AgentMarketplaceItem]
    total: int
    page: int
    page_size: int
